File: SZE/OODB/kyberszittya/Rec3/Recitation3/Labirintus/hero.cpp
#include "Hero.h"
#include "Padloelem.h"

Hero::Hero(const std::string name, unsigned int health, unsigned int attack, unsigned int damage, unsigned int defense)
	:
	Unit(health, attack, damage, defense),
	name(name)
{}

Hero::~Hero()
{
	current_padloelem = NULL;
}






std::ostream& operator<<(std::ostream& os, const Hero& hero)
{
	os << hero.name << hero.current_padloelem->getCoord();
	return os;
}

File: Misc/RPG_projekt/src/0220_healing/Hero.java
class Hero extends Unit {
    private String title = "";
    private int experience = 0;
    private int maxHealth;

    public Hero(String name, int health, int damage, int defense) {
        super(name, health, damage, defense);
        this.maxHealth = health;
    } 

    public void heal(int points) {
        if (isAlive()) {
            this.health += points;
            if (this.health > this.maxHealth) 
                this.health = this.maxHealth;
        }
    }

    public void healToFull() {
        if (isAlive()) this.health = this.maxHealth;
    }

    public void setTitle(String newTitle) {
        this.title = newTitle;
    }

    public String getOfficialName() {
        return this.title.equals("") ? this.name : this.name + ", the " + this.title;
    }

    @Override
    public void attack(Unit other) {
        super.attack(other);
        gainExperience();
    }

    private void gainExperience() {
        this.experience++;
        if (this.experience % 10 == 0) levelUp();
    }

    private void levelUp() {
        this.damage++;
        this.defense++;
        this.maxHealth += 10;
        healToFull();
    }    
}